File: Sample Code/ObjcSampleCode/Demo/DJISdkDemo/Demo/FlightController/FlightLimitation/FCFlightLimitationViewController.m
//
//  FCFlightLimitationViewController.m
//  DJISdkDemo
//
//  Copyright © 2015 DJI. All rights reserved.
//
/**
 *  This file demonstrates how to configure the flight limitaion setting through DJIFlightController.
 */
#import <DJISDK/DJISDK.h>
#import "DemoUtility.h"
#import "FCFlightLimitationViewController.h"

@interface FCFlightLimitationViewController () <UITextFieldDelegate>
@property (weak, nonatomic) IBOutlet UITextField *heightLimitTextField;
@property (weak, nonatomic) IBOutlet UITextField *radiusLimitTextField;
@property (weak, nonatomic) IBOutlet UISwitch *radiusLimitSwitch;
- (IBAction)onLimitSwitchValueChanged:(id)sender;

@end

@implementation FCFlightLimitationViewController

/**
 *  Display current limitation when view loaded.
 */
- (void)viewDidLoad {
    [super viewDidLoad];
    
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        // initialize the UI
        [self displayMaxFlightHeight];
        [self displayFlightRadiusLimitation];
    }
}

-(void) displayMaxFlightHeight {
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {        
        WeakRef(target);
        [fc.flightLimitation getMaxFlightHeightWithCompletion:^(float height, NSError * _Nullable error) {
            WeakReturn(target);
            if (error) {
                ShowResult(@"Get Max Flight Height:%@", error.localizedDescription);
            }
            else
            {
                target.heightLimitTextField.text = [NSString stringWithFormat:@"%0.1f", height];
            }
        }];
    }
}

/**
 *  The setting for flight radius limitation is a bit complicated than flight height limitation. SDK provides the capability to toggle
 *  the radius limitation. Therefore, before fetching the exact radius limitation, we check if it is already enabled or not.
 */
-(void) displayFlightRadiusLimitation {
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        WeakRef(target);
        [fc.flightLimitation getMaxFlightRadiusLimitationEnabledWithCompletion:^(BOOL enabled, NSError * _Nullable error) {
            WeakReturn(target);
            if (error) {
                ShowResult(@"Get RadiusLimitationEnable:%@", error.localizedDescription);
            }
            else
            {
                [target.radiusLimitSwitch setOn:enabled];
                
                // We check the flight radius limitation only if it is enabled.
                if (enabled) {
                    target.radiusLimitTextField.enabled = YES;
                    [fc.flightLimitation getMaxFlightRadiusWithCompletion:^(float radius, NSError * _Nullable error) {
                        WeakReturn(target);
                        if (error) {
                            ShowResult(@"Get MaxFlightRadius:%@", error.localizedDescription);
                        }
                        else
                        {
                            target.radiusLimitTextField.text = [NSString stringWithFormat:@"%0.1f", radius];
                        }
                    }];
                }
                else
                {
                    target.radiusLimitTextField.enabled = NO;
                    target.radiusLimitTextField.text = @"0";
                }
            }
        }];
    }
}

/**
 *  User can enable or disable the flight radius limitation using the UI component. 
 *  Whenever the functionality is enabled, we fetch the limitation and display it.
 */
- (IBAction)onLimitSwitchValueChanged:(UISwitch*)sender {
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        WeakRef(target);
        [fc.flightLimitation setMaxFlightRadiusLimitationEnabled:sender.on withCompletion:^(NSError * _Nullable error) {
            if (error) {
                ShowResult(@"setMaxFlightRadiusLimitationEnabled:%@", error.localizedDescription);
                [sender setOn:!sender.on animated:YES];
            }
            else
            {
                WeakReturn(target);
                if (sender.on) {
                    target.radiusLimitTextField.enabled = YES;
                    [fc.flightLimitation getMaxFlightRadiusWithCompletion:^(float radius, NSError * _Nullable error) {
                        WeakReturn(target);
                        if (error) {
                            ShowResult(@"%@", error.localizedDescription);
                        }
                        else
                        {
                            target.radiusLimitTextField.text = [NSString stringWithFormat:@"%0.1f", radius];
                        }
                    }];
                }
                else
                {
                    target.radiusLimitTextField.enabled = NO;
                    target.radiusLimitTextField.text = @"0";
                }
            }
        }];
    }
}

-(void) setMaxFlightHeight:(float)height
{
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        [fc.flightLimitation setMaxFlightHeight:height withCompletion:^(NSError * _Nullable error) {
            if (error) {
                ShowResult(@"setMaxFlightHeight:%@", error.localizedDescription);
            }
        }];
    }
}

-(void) setMaxFlightRadius:(float)radius
{
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        [fc.flightLimitation setMaxFlightRadius:radius withCompletion:^(NSError * _Nullable error) {
            if (error) {
                ShowResult(@"setMaxFlightRadius:%@", error.localizedDescription);
            }
        }];
    }
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    if ([textField isEqual:self.heightLimitTextField]) {
        float value = [textField.text floatValue];
        [self setMaxFlightHeight:value];
    }
    else
    {
        float value = [textField.text floatValue];
        [self setMaxFlightRadius:value];
    }
    
    if (textField.isFirstResponder) {
        [textField resignFirstResponder];
    }
    return YES;
}
@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/FlightController/FlightLimitation/FCFlightLimitationViewController.swift
//
//  FCFlightLimitationViewController.swift
//  DJISDKSwiftDemo
//
//  Created by DJI on 16/1/5.
//  Copyright © 2016 DJI. All rights reserved.
//

import DJISDK

class FCFlightLimitationViewController: DJIBaseViewController, DJIFlightControllerDelegate {
    @IBOutlet weak var heightLimitTextField: UITextField!
    @IBOutlet weak var radiusLimitTextField: UITextField!
    @IBOutlet weak var radiusLimitSwitch: UISwitch!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.getParameters()
        // Do any additional setup after loading the view from its nib.
      }
    
    func getParameters() {

        let fc: DJIFlightController? = self.fetchFlightController()
        if fc != nil {
            fc?.delegate = self
            
            fc?.flightLimitation?.getMaxFlightHeight(completion: {[weak self](height: Float, error: Error?) -> Void in
                if error != nil {
                    self?.showAlertResult("Get Max Flight Height:\(error?.localizedDescription)")
                }
                else {
                    self?.heightLimitTextField.text = String(format: "%0.1f", height)
                }
            })
            
            fc?.flightLimitation?.getMaxFlightRadiusLimitationEnabled(completion: {[weak self](enabled: Bool, error: Error?) -> Void in
                
                if error != nil {
                    self?.showAlertResult("Get RadiusLimitationEnable:\(error?.localizedDescription)")
                }
                else {
                    self?.radiusLimitSwitch.isOn = enabled
                    if enabled != false {
                        self?.radiusLimitTextField.isEnabled = true
                        fc?.flightLimitation?.getMaxFlightRadius(completion: {[weak self](radius: Float, error: Error?) -> Void in
                            
                            if error != nil {
                                self?.showAlertResult("Get MaxFlightRadius:\(error?.localizedDescription)")
                            }
                            else {
                                self?.radiusLimitTextField.text = String(format: "%0.1f", radius)
                            }
                            
                            })
                    }
                    else {
                        self?.radiusLimitTextField.isEnabled = false
                        self?.radiusLimitTextField.text = "0"
                    }
                }
                })
        }

    
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func onLimitSwitchValueChanged(_ sender: UISwitch) {
        let fc: DJIFlightController? = self.fetchFlightController()
        if fc != nil {
            fc?.flightLimitation?.setMaxFlightRadiusLimitationEnabled(sender.isOn, withCompletion: {[weak self](error: Error?) -> Void in
                if error != nil {
                    self?.showAlertResult("setMaxFlightRadiusLimitationEnabled:\(error?.localizedDescription)")
                    sender.setOn(!sender.isOn, animated: true)
                }
                else {
                        if sender.isOn {
                            self?.radiusLimitTextField.isEnabled = true
                            fc?.flightLimitation?.getMaxFlightRadius(completion: {[weak self](radius: Float, error: Error?) -> Void in
                                if error != nil {
                                    self?.showAlertResult("\(error?.localizedDescription)")
                                }
                                else {
                                        self?.radiusLimitTextField.text = String(format: "%0.1f", radius)
                                }
                                
                                fc?.flightLimitation?.getMaxFlightHeight(completion: {[weak self](height: Float, error: Error?) -> Void in
                                    if error != nil {
                                        self?.showAlertResult("Get Max Flight Height:\(error?.localizedDescription)")
                                    }
                                    else {
                                        self?.heightLimitTextField.text = String(format: "%0.1f", height)
                                    }
                                    })
                            })
                            
                        }
                        else {
                            self?.radiusLimitTextField.isEnabled = false
                            self?.radiusLimitTextField.text = "0"
                        }
                }
            })
        }
    }
    
    func setMaxFlightHeight(_ height: Float) {
        let fc: DJIFlightController? = self.fetchFlightController()
        if fc != nil {
            fc?.flightLimitation?.setMaxFlightHeight(height, withCompletion: {[weak self](error: Error?) -> Void in
                //if error != nil {
                    self?.showAlertResult("setMaxFlightHeight:\(error?.localizedDescription)")
                //}
            })
        }
    }
    
    func setMaxFlightRadius(_ radius: Float) {
        let fc: DJIFlightController? = self.fetchFlightController()
        if fc != nil {
            fc?.flightLimitation?.setMaxFlightRadius(radius, withCompletion: {[weak self](error: Error?) -> Void in
                //if error != nil {
                    self?.showAlertResult("setMaxFlightRadius:\(error?.localizedDescription)")
                //}
            })
        }
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if textField.isEqual(self.heightLimitTextField) {
            let value: Float = CFloat(textField.text!)!
            self.setMaxFlightHeight(value)
        }
        else {
            let value: Float = CFloat(textField.text!)!
            self.setMaxFlightRadius(value)
        }
        if textField.isFirstResponder as Bool == true {
            textField.resignFirstResponder()
        }
        return true
    }
    
 
}
